packages/langium/src/grammar/type-system/type-collector/types.ts

```diff
@@ -304,7 +304,18 @@ function isTypeAssignableInternal(from: PropertyType | undefined, to: PropertyTy
         result = isArrayType(to) && isTypeAssignableInternal(from.elementType, to.elementType, visited);
     } else if (isValueType(from)) {
         if (isUnionType(from.value)) {
-            result = isTypeAssignableInternal(from.value.type, to, visited);
+            if (from.value.dataType) {
+                // We can test the primitive data type directly
+                // This potentially skips a expensive recursive call
+                // This also helps in case the computed internal data type does not fit the declared data type
+                const primitiveType: PrimitiveType = {
+                    primitive: from.value.dataType
+                };
+                result = isTypeAssignableInternal(primitiveType, to, visited);
```

Comment:
Overall this looks okay for the number -> string case. But the other way around doesn't feel right. If I flip around the test below so it converts from string -> number, it also passes, but I wouldn't expect that we can assign an arbitrary string to a number type.

I figured we're doing a type check on the Datatype to make sure it's computed type is reasonable given the declared type, but it seems like this is a gap somehow.

```antlr
interface RuleType { prop : number };
Rule returns RuleType: prop = MyDataType;
MyDataType returns number: ID;
terminal ID returns string: /[_a-zA-Z][\\w_]*/;
```

Reply:
Eventually, this simply runs through the value converter and attempts to do `Number(input)`. This will likely result in `NaN`, but it's a "valid" conversion.

I agree that it doesn't make a lot of sense, but if users want to shoot themselves in the foot, who are we to stop them? :D

Reply:
Yeah I saw that too. I still feel like maybe we can provide some advice about avoiding this at the least, maybe in a quite write-up or something later on. But that's a separate thing then.